File: src/main/java/lekkit/scev/server/FirmwareAssets.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.server;

import com.mojang.logging.LogUtils;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.Objects;
import java.util.zip.CRC32;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;

/**
 * Ships read-only firmware/OS blobs from the mod jar into {@code ./scev/assets/}
 * on first use. This is the backbone of the "out-of-box" experience: a freshly
 * installed mod boots real OpenSBI + U-Boot instead of a 16-byte demo loop,
 * because this class materializes {@code fw_payload.bin} (and any other bundled
 * blob) before {@link StorageManager#copyImage} goes looking for it.
 *
 * <h2>Asset lookup priority</h2>
 * <ol>
 *   <li><b>Classpath (bundled)</b> — {@code /assets/scev/firmware/&lt;name&gt;} inside
 *       the mod jar. Extracted to disk the first time the asset is requested.</li>
 *   <li><b>Local disk (user-supplied)</b> — {@code ./scev/assets/&lt;name&gt;}. If
 *       the user drops their own firmware/kernel here, we use it in preference
 *       to extracting the bundled copy <i>only</i> when no bundled copy exists.
 *       Users who want to override the bundled copy can delete
 *       {@code ./scev/assets/&lt;name&gt;} <i>and</i> set {@code scev.firmware.force_bundled=false}.</li>
 * </ol>
 *
 * <h2>Extraction semantics</h2>
 * <ul>
 *   <li><b>Idempotent:</b> calling {@link #ensureExtracted} twice is a no-op on
 *       the second call. File existence is checked up front.</li>
 *   <li><b>Atomic:</b> extraction writes to a temp file under the same
 *       directory then renames — no readers ever see a half-written file.</li>
 *   <li><b>Fail-soft:</b> extraction failure logs a warning and returns the
 *       on-disk path anyway. Callers must still {@link #isAvailable check}
 *       before using.</li>
 * </ul>
 *
 * <h2>Why a separate class (not folded into {@link StorageManager})</h2>
 * <ul>
 *   <li>Firmware assets are read-only, fixed-size, shared across all machines.
 *       Disk images are per-UUID, mutable, lazy-created.</li>
 *   <li>Tests for extraction want to point at an ephemeral directory without
 *       touching {@code ./scev/images/}.</li>
 *   <li>Lets {@link StorageManager#copyImage} stay a pure file-copy helper with
 *       no knowledge of the classpath.</li>
 * </ul>
 */
public final class FirmwareAssets {
    private static final Logger LOG = LogUtils.getLogger();

    /** Classpath prefix for bundled firmware resources. */
    public static final String CLASSPATH_PREFIX = "/assets/scev/firmware/";

    /** On-disk directory where extracted + user-supplied assets live. */
    private static final Path ASSETS_DIR = Paths.get("scev", "assets");

    /**
     * Canonical name of the default M-mode firmware blob (OpenSBI + U-Boot).
     * Matches {@link lekkit.scev.items.FlashItem}'s declared origin.
     */
    public static final String DEFAULT_FIRMWARE = "fw_payload.bin";

    static {
        try {
            Files.createDirectories(ASSETS_DIR);
        } catch (IOException ignore) {
            // Will be lazily created on first extraction attempt.
        }
    }

    private FirmwareAssets() {}

    /**
     * Is there a classpath-bundled copy of {@code name}? This is decided at
     * runtime by peeking at the mod jar's resource table — so removing a
     * firmware file from the jar immediately switches any caller to the
     * user-supplied path without any code changes.
     */
    public static boolean isBundled(String name) {
        Objects.requireNonNull(name, "name");
        return FirmwareAssets.class.getResource(CLASSPATH_PREFIX + name) != null;
    }

    /**
     * Is the asset available for use — either bundled in the jar or already
     * extracted/placed on disk. Callers use this as a precondition before
     * passing the path to RVVM's {@code load_bootrom} / {@code load_kernel}.
     */
    public static boolean isAvailable(String name) {
        Objects.requireNonNull(name, "name");
        return isBundled(name) || Files.isRegularFile(ASSETS_DIR.resolve(name));
    }

    /**
     * The on-disk path an asset <i>would</i> live at if present. Does not
     * create or extract anything. Useful for building paths that will be
     * populated later.
     */
    public static Path diskPath(String name) {
        Objects.requireNonNull(name, "name");
        return ASSETS_DIR.resolve(name);
    }

    /**
     * Ensure the named asset exists on disk and return its absolute path.
     *
     * <p>Priority logic:
     * <ul>
     *   <li>Extracted file's CRC32 matches the bundled copy's CRC32 → use the
     *       extract (cheap cache hit, avoids re-copying the ~65 MB Alpine
     *       image each boot).</li>
     *   <li>Extracted file exists but bundled copy has a different CRC → the
     *       mod's bundled asset was updated (e.g. new kernel shipped); the
     *       extract is stale. Re-extract and warn.</li>
     *   <li>Extracted file exists and no bundled copy on the classpath →
     *       user-supplied fallback; trust it and use as-is.</li>
     *   <li>No extract and bundled copy present → extract.</li>
     *   <li>Neither → {@code null}, caller handles.</li>
     * </ul>
     *
     * <p>History: the original implementation did "extract-wins-silently," so
     * stale extracts from older mod versions kept loading forever. Size
     * comparison caught most real upgrades but silently produced stale bytes
     * when two different builds happened to be the same size (e.g. two
     * revisions of a 65 MiB partitioned ext4 image). Switched to CRC32 after
     * an image-contents-differ-but-size-identical bug in the Alpine pipeline.
     * The CRC stream costs ~200 ms on a 65 MiB asset per boot — acceptable,
     * runs once-per-asset-per-process, and rules out silent collisions.
     *
     * <p>Thread-safe: synchronized on the class so concurrent extraction
     * attempts don't race.
     */
    public static synchronized @Nullable Path ensureExtracted(String name) {
        Objects.requireNonNull(name, "name");
        Path target = ASSETS_DIR.resolve(name);
        boolean targetExists = Files.isRegularFile(target);
        boolean bundled = isBundled(name);

        if (targetExists && !bundled) {
            // User-supplied fallback — no bundled copy to compare against, so
            // we trust the on-disk file completely.
            return target.toAbsolutePath();
        }
        if (targetExists && bundled) {
            long bundledCrc = bundledCrcOrMinusOne(name);
            long targetCrc = fileCrcOrMinusOne(target);
            if (bundledCrc >= 0 && targetCrc >= 0 && targetCrc == bundledCrc) {
                // Cache hit — extracted copy's CRC matches the bundled copy.
                return target.toAbsolutePath();
            }
            LOG.warn("Firmware asset {} on disk (CRC={}) differs from bundled copy (CRC={}); "
                            + "re-extracting. If you intentionally customised this file, rename "
                            + "it before the mod upgrade clobbers it again.",
                    name, String.format("%08x", targetCrc), String.format("%08x", bundledCrc));
            // Fall through to re-extract below.
        }
        if (!bundled) {
            LOG.debug("Firmware asset {} not bundled and not on disk", name);
            return null;
        }
        try {
            Files.createDirectories(ASSETS_DIR);
        } catch (IOException e) {
            LOG.warn("Could not create {} for firmware assets", ASSETS_DIR, e);
            return null;
        }
        // Write atomically: temp file + move. A crashed extraction never
        // leaves a half-written asset that later callers would silently read.
        Path tmp;
        try {
            tmp = Files.createTempFile(ASSETS_DIR, name + ".", ".part");
        } catch (IOException e) {
            LOG.warn("Could not create temp file under {}", ASSETS_DIR, e);
            return null;
        }
        try (InputStream in = FirmwareAssets.class.getResourceAsStream(CLASSPATH_PREFIX + name);
             OutputStream out = Files.newOutputStream(tmp)) {
            if (in == null) {
                LOG.warn("Bundled firmware asset {} disappeared between probe and open", name);
                Files.deleteIfExists(tmp);
                return null;
            }
            in.transferTo(out);
        } catch (IOException e) {
            LOG.warn("Failed to extract bundled firmware asset {}", name, e);
            try { Files.deleteIfExists(tmp); } catch (IOException ignore) {}
            return null;
        }
        try {
            Files.move(tmp, target, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException atomic) {
            // Fallback for filesystems that don't support ATOMIC_MOVE (e.g. across mounts).
            try {
                Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException fatal) {
                LOG.warn("Failed to place extracted firmware asset {} at {}", name, target, fatal);
                try { Files.deleteIfExists(tmp); } catch (IOException ignore) {}
                return null;
            }
        }
        LOG.info("Extracted bundled firmware asset {} to {}", name, target);
        return target.toAbsolutePath();
    }

    /**
     * Return the absolute path of an extracted asset, or {@code null} if it
     * can't be made available (neither bundled nor on disk). Convenience for
     * call sites that want the path directly.
     */
    public static @Nullable String resolveAbsolutePath(String name) {
        Path p = ensureExtracted(name);
        return p == null ? null : p.toString();
    }

    /** Size of {@code target} in bytes, or -1 if it can't be stat'd. */
    private static long sizeOrMinusOne(Path target) {
        try {
            return Files.size(target);
        } catch (IOException e) {
            return -1;
        }
    }

    /** CRC32 of the on-disk file, or -1 if it can't be read. */
    private static long fileCrcOrMinusOne(Path target) {
        try (InputStream in = Files.newInputStream(target)) {
            return streamCrc(in);
        } catch (IOException e) {
            return -1;
        }
    }

    /**
     * CRC32 of the classpath-bundled asset, or -1 if it isn't bundled or
     * can't be read. Unlike {@link #bundledSizeOrMinusOne} this distinguishes
     * two bundled assets of identical size but different contents — the
     * point of switching from size to CRC.
     */
    private static long bundledCrcOrMinusOne(String name) {
        try (InputStream in = FirmwareAssets.class.getResourceAsStream(CLASSPATH_PREFIX + name)) {
            if (in == null) return -1;
            return streamCrc(in);
        } catch (IOException e) {
            return -1;
        }
    }

    /**
     * Streaming CRC32 — hashes an input stream in 16 KiB chunks. Rolled
     * manually because {@link java.util.zip.CheckedInputStream} would add
     * an extra layer and callers already own the stream lifecycle.
     */
    private static long streamCrc(InputStream in) throws IOException {
        CRC32 crc = new CRC32();
        byte[] buf = new byte[16 * 1024];
        int n;
        while ((n = in.read(buf)) > 0) crc.update(buf, 0, n);
        return crc.getValue();
    }

    /**
     * Size in bytes of the classpath-bundled asset {@code name}, or -1 if it
     * isn't bundled or can't be read. We stream the resource to count bytes —
     * classpath resources don't expose a size attribute through the URL API.
     */
    private static long bundledSizeOrMinusOne(String name) {
        try (InputStream in = FirmwareAssets.class.getResourceAsStream(CLASSPATH_PREFIX + name)) {
            if (in == null) return -1;
            long total = 0;
            byte[] buf = new byte[16 * 1024];
            int n;
            while ((n = in.read(buf)) > 0) total += n;
            return total;
        } catch (IOException e) {
            return -1;
        }
    }

    /**
     * Size in bytes of the on-disk asset (after extraction if needed).
     * Returns -1 when the asset isn't available. Used by tests and by
     * diagnostics so we can report "firmware loaded, size=2.7 MiB" in logs.
     */
    public static long sizeBytes(String name) {
        Path p = ensureExtracted(name);
        if (p == null) return -1;
        try {
            return Files.size(p);
        } catch (IOException e) {
            return -1;
        }
    }

    /**
     * Reset extraction state — for tests only. Deletes the on-disk copy if it
     * exists. Does nothing to the bundled resource.
     */
    public static void forgetExtracted(String name) {
        try {
            Files.deleteIfExists(ASSETS_DIR.resolve(name));
        } catch (IOException ignore) {
        }
    }
}

File: src/main/kotlin/lekkit/scev/server/FirmwareAssets.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package lekkit.scev.server

import com.mojang.logging.LogUtils
import java.io.IOException
import java.io.InputStream
import java.nio.ByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.util.zip.CRC32

/**
 * Ships read-only firmware/OS blobs from the mod jar into `./scev/assets/`
 * on first use. Backbone of the "out-of-box" experience: a freshly installed
 * mod boots real OpenSBI + U-Boot instead of a 16-byte demo loop because
 * this class materialises `fw_payload.bin` (and any other bundled blob)
 * before [StorageManager.copyImage] goes looking for it.
 *
 * **Asset lookup priority.**
 *  1. **Classpath (bundled)** — `/assets/scev/firmware/<name>` inside the
 *     mod jar. Extracted to disk on first request.
 *  2. **Local disk (user-supplied)** — `./scev/assets/<name>`. If the user
 *     drops their own firmware/kernel, we use it in preference to extracting
 *     the bundled copy *only* when no bundled copy exists. To intentionally
 *     override the bundled copy, set `scev.firmware.force_bundled=false`
 *     and place the file at `./scev/assets/<name>`.
 *
 * **Extraction semantics.**
 *  - Idempotent: [ensureExtracted] is a no-op on subsequent calls.
 *  - Atomic: writes to a temp file under the same dir, then renames.
 *  - Fail-soft: extraction failure logs a warning and returns `null`;
 *    callers must check before using.
 */
object FirmwareAssets {
    private val LOG = LogUtils.getLogger()

    /** Classpath prefix for bundled firmware resources. */
    @JvmField val CLASSPATH_PREFIX: String = "/assets/scev/firmware/"

    /** On-disk directory where extracted + user-supplied assets live. */
    private val ASSETS_DIR: Path = Paths.get("scev", "assets")

    /**
     * Canonical name of the default M-mode firmware blob (OpenSBI + U-Boot).
     * Matches `FlashItem`'s declared origin.
     */
    @JvmField val DEFAULT_FIRMWARE: String = "fw_payload.bin"

    init {
        try { Files.createDirectories(ASSETS_DIR) } catch (_: IOException) {
            // Will be lazily created on first extraction attempt.
        }
    }

    /**
     * Is there a classpath-bundled copy of [name]? Decided at runtime by
     * peeking at the mod jar's resource table — removing a firmware file
     * from the jar immediately switches any caller to the user-supplied
     * path without code changes.
     */
    @JvmStatic fun isBundled(name: String): Boolean =
        FirmwareAssets::class.java.getResource(CLASSPATH_PREFIX + name) != null

    /**
     * Is the asset available — bundled in the jar or already on disk.
     * Callers use this as a precondition before passing the path to RVVM's
     * `load_bootrom` / `load_kernel`.
     */
    @JvmStatic fun isAvailable(name: String): Boolean =
        isBundled(name) || Files.isRegularFile(ASSETS_DIR.resolve(name))

    /**
     * The on-disk path an asset *would* live at if present. Does not create
     * or extract anything. Useful for building paths populated later.
     */
    @JvmStatic fun diskPath(name: String): Path = ASSETS_DIR.resolve(name)

    /**
     * Ensure the named asset exists on disk and return its absolute path.
     *
     * Priority:
     *  - Extracted file's CRC32 matches bundled copy → cache hit, return path.
     *  - Extracted exists but bundled CRC differs → mod's bundled asset was
     *    updated; the extract is stale. Re-extract and warn.
     *  - Extracted exists, no bundled copy → user-supplied fallback. Trust it.
     *  - No extract, bundled present → extract.
     *  - Neither → null, caller handles.
     *
     * Switched from size comparison to CRC32 after an image-contents-differ-
     * but-size-identical bug in the Alpine pipeline. CRC stream costs ~200 ms
     * on a 65 MiB asset per boot — acceptable, runs once-per-asset-per-process.
     *
     * Thread-safe: synchronized so concurrent extraction attempts don't race.
     */
    @JvmStatic @Synchronized fun ensureExtracted(name: String): Path? {
        val target = ASSETS_DIR.resolve(name)
        val targetExists = Files.isRegularFile(target)
        val bundled = isBundled(name)

        if (targetExists && !bundled) {
            // User-supplied fallback — no bundled copy to compare against.
            return target.toAbsolutePath()
        }
        if (targetExists && bundled) {
            val bundledCrc = bundledCrcOrMinusOne(name)
            val targetCrc = fileCrcOrMinusOne(target)
            if (bundledCrc >= 0 && targetCrc >= 0 && targetCrc == bundledCrc) {
                return target.toAbsolutePath()  // CRC match — cache hit
            }
            LOG.warn("Firmware asset {} on disk (CRC={}) differs from bundled copy (CRC={}); " +
                "re-extracting. If you intentionally customised this file, rename it before " +
                "the mod upgrade clobbers it again.",
                name, "%08x".format(targetCrc), "%08x".format(bundledCrc))
            // Fall through to re-extract.
        }
        if (!bundled) {
            LOG.debug("Firmware asset {} not bundled and not on disk", name)
            return null
        }
        try {
            Files.createDirectories(ASSETS_DIR)
        } catch (e: IOException) {
            LOG.warn("Could not create {} for firmware assets", ASSETS_DIR, e)
            return null
        }
        // Write atomically: temp file + move. A crashed extraction never
        // leaves a half-written asset that later callers would silently read.
        val tmp = try {
            Files.createTempFile(ASSETS_DIR, "$name.", ".part")
        } catch (e: IOException) {
            LOG.warn("Could not create temp file under {}", ASSETS_DIR, e)
            return null
        }

        // Sparse-aware extraction — see sparseStreamCopy. The shipped images
        // are mostly-empty ext4; a plain transferTo would write every zero
        // block, costing ~1 GiB for a 1 GiB image that's only ~70 MiB of
        // real data.
        try {
            FirmwareAssets::class.java.getResourceAsStream(CLASSPATH_PREFIX + name).use { input ->
                FileChannel.open(tmp, StandardOpenOption.WRITE, StandardOpenOption.SPARSE).use { out ->
                    if (input == null) {
                        LOG.warn("Bundled firmware asset {} disappeared between probe and open", name)
                        Files.deleteIfExists(tmp)
                        return null
                    }
                    sparseStreamCopy(input, out)
                }
            }
        } catch (e: IOException) {
            LOG.warn("Failed to extract bundled firmware asset {}", name, e)
            try { Files.deleteIfExists(tmp) } catch (_: IOException) {}
            return null
        }

        try {
            Files.move(tmp, target, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        } catch (_: IOException) {
            // Fallback for filesystems that don't support ATOMIC_MOVE (across mounts).
            try {
                Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING)
            } catch (fatal: IOException) {
                LOG.warn("Failed to place extracted firmware asset {} at {}", name, target, fatal)
                try { Files.deleteIfExists(tmp) } catch (_: IOException) {}
                return null
            }
        }
        LOG.info("Extracted bundled firmware asset {} to {}", name, target)
        return target.toAbsolutePath()
    }

    /**
     * Absolute path of the extracted asset, or null if it can't be made
     * available. Convenience for call sites that want the path directly.
     */
    @JvmStatic fun resolveAbsolutePath(name: String): String? =
        ensureExtracted(name)?.toString()

    /** Size in bytes of the on-disk asset (extracts if needed), or -1. */
    @JvmStatic fun sizeBytes(name: String): Long {
        val p = ensureExtracted(name) ?: return -1
        return try { Files.size(p) } catch (_: IOException) { -1 }
    }

    /** Reset extraction state — tests only. Deletes the on-disk copy. */
    @JvmStatic fun forgetExtracted(name: String) {
        try { Files.deleteIfExists(ASSETS_DIR.resolve(name)) } catch (_: IOException) {}
    }

    /* ----- Internals ------------------------------------------------------ */

    /** CRC32 of the on-disk file, or -1 if unreadable. */
    private fun fileCrcOrMinusOne(target: Path): Long = try {
        Files.newInputStream(target).use { streamCrc(it) }
    } catch (_: IOException) { -1 }

    /** CRC32 of the bundled asset, or -1 if not bundled / unreadable. */
    private fun bundledCrcOrMinusOne(name: String): Long = try {
        FirmwareAssets::class.java.getResourceAsStream(CLASSPATH_PREFIX + name)
            ?.use { streamCrc(it) } ?: -1L
    } catch (_: IOException) { -1L }

    /**
     * Stream-to-FileChannel copy that preserves sparseness by skipping 64 KiB
     * zero-filled chunks. Destination [FileChannel.position] advances past
     * each hole — on sparse-capable filesystems (APFS/ext4/NTFS) the skipped
     * region is never allocated.
     *
     * Final file size is the sum of bytes read; when the tail is a zero
     * chunk we extend the file by writing a single zero byte at the end.
     *
     * 64 KiB chunk: amortises syscall cost; small enough that non-zero
     * content surrounded by zeros doesn't pull a whole MiB into allocation.
     */
    private fun sparseStreamCopy(input: InputStream, out: FileChannel) {
        val chunkSize = 64 * 1024
        val buf = ByteArray(chunkSize)
        var pos = 0L
        var lastChunkSkipped = false
        while (true) {
            val filled = fillBuffer(input, buf)
            if (filled == 0) break
            if (isAllZero(buf, filled)) {
                pos += filled
                lastChunkSkipped = true
                continue
            }
            out.position(pos)
            val bb = ByteBuffer.wrap(buf, 0, filled)
            var written = 0
            while (written < filled) written += out.write(bb)
            pos += filled
            lastChunkSkipped = false
        }
        // Preserve logical length when the tail was all zeros: write a
        // single zero byte at pos-1 so the file shows the right size.
        if (lastChunkSkipped && pos > 0) {
            out.position(pos - 1)
            out.write(ByteBuffer.wrap(byteArrayOf(0)))
        }
    }

    /**
     * Read as close to [buf].size bytes as the stream provides. Returns the
     * total read. Works around `InputStream.read(byte[])` returning early on
     * a partial buffer even when more data is available — which would make
     * zero-chunk detection misalign from the 64 KiB grid.
     */
    private fun fillBuffer(input: InputStream, buf: ByteArray): Int {
        var total = 0
        while (total < buf.size) {
            val n = input.read(buf, total, buf.size - total)
            if (n < 0) break
            total += n
        }
        return total
    }

    /** True iff the first [len] bytes of [buf] are all zero. */
    private fun isAllZero(buf: ByteArray, len: Int): Boolean {
        for (i in 0 until len) if (buf[i] != 0.toByte()) return false
        return true
    }

    /**
     * Streaming CRC32 — hashes a stream in 16 KiB chunks. Rolled manually
     * because [java.util.zip.CheckedInputStream] would add a wrapper layer
     * and callers already own the stream lifecycle.
     */
    private fun streamCrc(input: InputStream): Long {
        val crc = CRC32()
        val buf = ByteArray(16 * 1024)
        while (true) {
            val n = input.read(buf)
            if (n <= 0) break
            crc.update(buf, 0, n)
        }
        return crc.value
    }
}
